web-api/src/storage/elastic.rs

```diff
@@ -145,72 +150,68 @@ impl Client {
-                        properties: &document.properties,
-                        // TODO[pmk/ET-4756-7]
-                        embedding: &document.snippets[0].1,
-                        tags: &document.tags,
-                    })
-                    .map_err(Into::into),
-                ]
-            }))
-            .await?;
-        Ok(response.failed_documents("index", false).into())
+        let ids = documents.iter().map(|document| &document.id).collect_vec();
+        self.delete_by_parents(ids).await?;
+        self.freshly_insert_documents(documents).await
     }
 
-    pub(super) async fn delete_documents(
+    pub(super) async fn freshly_insert_documents(
         &self,
-        ids: impl IntoIterator<IntoIter = impl ExactSizeIterator<Item = &DocumentId>>,
+        documents: impl IntoIterator<Item = &models::DocumentForIngestion>,
     ) -> Result<Warning<DocumentId>, Error> {
-        let ids = ids.into_iter();
-        if ids.len() == 0 {
+        // TODO[pmk/now] this won't work correctly with upserts id old snippet count > new snippet count
+        let mut snippets = documents
+            .into_iter()
+            .flat_map(|document| {
+                document
+                    .snippets
+                    .iter()
+                    .enumerate()
+                    .flat_map(|(idx, (snippet, embedding))| {
+                        #[allow(clippy::cast_possible_truncation)]
+                        let id = SnippetId::new(document.id.clone(), idx as _);
+                        let header =
+                            serde_json::to_value(BulkInstruction::Create { id: id.to_es_id() })
+                                .map_err(Into::into);
+                        let data = serde_json::to_value(EsDocument {
+                            snippet,
+                            properties: &document.properties,
+                            embedding,
+                            tags: &document.tags,
+                            parent: id.document_id(),
+                        })
+                        .map_err(Into::into);
+                        [header, data]
+                    })
+            })
+            .peekable();
+
+        if snippets.peek().is_none() {
             return Ok(Warning::default());
         }
 
-        let response = self
-            .bulk_request(ids.map(|id| Ok(BulkInstruction::Delete { id })))
-            .await?;
-        Ok(response.failed_documents("delete", true).into())
+        let response = self.bulk_request(snippets).await?;
+        Ok(response.failed_documents("index", false).into())
     }
 
-    pub(super) async fn retain_documents(
+    pub(super) async fn delete_by_parents(
         &self,
-        ids: impl IntoIterator<IntoIter = impl ExactSizeIterator<Item = &DocumentId>>,
+        parents: impl SerializeDocumentIds,
     ) -> Result<(), Error> {
-        // https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-delete-by-query.html
```

Comment:
:art: we could keep the link to the es endpoint docs for easier access
```suggestion
        // https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-delete-by-query.html
```